pyccl/core.py

```diff
@@ -229,27 +229,53 @@ def write_yaml(self, filename):
         """Write a YAML representation of the parameters to file.
 
         Args:
-            filename (:obj:`str`) Filename to write parameters to.
+            filename (:obj:`str`) Filename, file pointer, or stream to write "
+                "parameters to."
         """
-        # NOTE: we use the C yaml dump here so that the parameters
-        # dumped by this object are compatible with the C yaml load function.
-        status = 0
-        status = lib.parameters_write_yaml(self._params, filename, status)
-
-        # Check status
-        if status != 0:
-            raise IOError("Unable to write YAML file {}".format(filename))
+        def make_yaml_friendly(d):
+            for k, v in d.items():
+                if isinstance(v, np.floating):
+                    d[k] = float(v)
+                elif isinstance(v, np.integer):
+                    d[k] = int(v)
+                elif isinstance(v, np.bool):
+                    d[k] = bool(v)
+                elif isinstance(v, dict):
+                    make_yaml_friendly(v)
+
+        params = self._params_init_kwargs.copy()
+        make_yaml_friendly(params)
+
+        if isinstance(filename, str):
+            with open(filename, "w") as fp:
+                yaml.dump(params, fp, default_flow_style=False)
+        else:
+            yaml.dump(params, filename, default_flow_style=False)
 
     @classmethod
     def read_yaml(cls, filename, **kwargs):
         """Read the parameters from a YAML file.
 
         Args:
-            filename (:obj:`str`) Filename to read parameters from.
+            filename (:obj:`str`) Filename, file pointer, or stream to read
+                parameters from.
             **kwargs (dict) Additional keywords that supersede file contents
         """
-        with open(filename, 'r') as fp:
-            params = yaml.load(fp, Loader=yaml.Loader)
+        if isinstance(filename, str):
+            with open(filename, 'r') as fp:
+                params = yaml.load(fp, Loader=yaml.Loader)
+        else:
+            params = yaml.load(filename, Loader=yaml.Loader)
+
+        # Check for missing values in yaml files created with the C-level
+        # yaml dumper
+        if any([p not in params for p in ["m_nu", "m_nu_type"]]):
```

Comment:
Is this the only case that needs to be checked?

Reply:
Technically `z_mg` and `df_mg` as well but I don't think that had ever worked (C-level calls them `z_mgrowth`). The new MG parameters (`c1_mg`, etc) don't have a fall-back either, so the readers already fail on yaml files created with old versions of CCL. 
So I'm tempted to replace that check with just a check for a `KeyError` and give a warning if any parameters are missing.

Reply:
The current version should handle this more cleanly and robustly.